### ISSUE TITLE
Feat: prevent lualine from overflowing the window width

### ISSUE THREAD

**Issue by hoob3rt** [hoob3rt/lualine.nvim#225]
### Requested feature

Add a global(or local) `config.shorten` option which will handle automatic shortening for all? default components.

### Motivation

As @shadmansaleh mentioned in #224 this can be done for each component with the `format` option. However this requires configuration bloat (imo) so I think we should provide sane shortening defaults to try preventing lualine from overflowing the window width.

**Comment by wookayin** [comment]
I think we can first start having a shortening feature (like `airline#util#shorten`) without multi-byte characters into much consideration. Once this is done, extending it to be aware of multi-byte characters wouldn't be very difficult.

**Comment by shadmansaleh** [comment]
What does `airline#util#shorten` do ? Is it expossed to users ? It feels like a internal helper function.

**Comment by wookayin** [comment]
@shadmansaleh Yes `airline#util#shorten(text, winwidth, minwidth)` is more like an internal function, but  which nvim-lualine will also likely to implement a similar thing as. It gives a truncated string including `...` if the current window width is less than `winwidth`, but having at least (either the first or the last)`minwidth` characters.

In addition to this (shorten/truncate with dots), airline controls whether some component is **visible or not** depending on the current window width:
```
" Here is how you would define a part that is visible only if the window width greater than a minimum width.
call airline#parts#define_minwidth('some_component', 50)
```
Similarily, for lualine we could have some configuration options, for example,
```
{
    'mode',
    ...
    min_width = ...;       -- If the current window's width is less then this threshold, hide the component
    shortening_target = ...;         -- Similar to that for 'filename', but more generally for all components
}
```

We can have either one of these, or probably both,

**Comment by shadmansaleh** [comment]
These can already be achieved with `fmt` 2nd one can be done with  `cond` .

For example:
```lua
local function trunc(trunc_width, trunc_len, hide_width, no_ellipsis)
  return function(str)
    local win_width = vim.fn.winwidth(0)
    if hide_width and win_width < hide_width then
       return ''
    elseif trunc_width and trunc_len and win_width < trunc_width and #str > trunc_len then
       return str:sub(1, trunc_len) .. (no_ellipsis and '' or '...')
    end
    return str
  end
end
 
...
lualine_a = {
    {'mode', fmt=trunc(80, 4, nil, true)},
    {'filename', fmt=trunc(120, 30, 50)},
    {function() return require'lsp-status'.status() end, fmt=truc(120, 20, 60)}
  }
```

The issue is this method won't work for [vim-stl-components](https://github.com/nvim-lualine/lualine.nvim#vims-statusline-items-as-lualine-component) and probably be broken with unicode characters.

**Comment by wookayin** [comment]
Thanks for sharing the snippet for how to truncate long components. I wish this were a built-in feature where a reasonable default behavior of truncating, but heavy users can start from those configs for more flexibility and customization.

In addition, how about conditional visibility? Is it something that should be implemented with `fmt` where a function simply returns an empty string?

**Comment by asteroidalaz** [comment]
A function which could tell if the lualine if overflowing?? Is that function available?

**Comment by shadmansaleh** [comment]
No currently there aren't any.

**Comment by rofrol** [comment]
It should be `vim.o.columns` instead of `vim.fn.winwidth(0)`. Otherwise when I open neo-tree.nvim, it gets winwidth from neo-tree window.
